src/core/dnd/types.ts

```diff
@@ -59,76 +58,94 @@ export const canDropIntoLayout = (
-  if (item.schema) {
-    // check wether the control fits to the scope
-    // -- TODO check other cases than array
-    const scopeSchemaElement = getArrayContainer(item.schema);
-    if (!!scopeSchemaElement !== !!scopeUISchemaElement) {
-      // scopes don't match when one exist and the other doesn't
-      return false;
-    }
-    if (
-      scopeSchemaElement &&
-      scopeUISchemaElement &&
-      scopeUISchemaElement.linkedSchemaElement !== scopeSchemaElement.uuid
-    ) {
-      // scopes didn't match
+  const controlObject = item.schema;
+  if (controlObject) {
+    const scopeSchemaElement = getScopeChangingContainer(controlObject);
+    if (!scopesMatch(scopeSchemaElement, scopeUISchemaElement)) {
       return false;
     }
   }
   return true;
 };
 
-export const canMoveSchemaElementTo = (
-  item: MoveUISchemaElement,
-  layout: EditorUISchemaElement,
-  index: number
+/**
+ * Scopes match if they are linked or both don't exist.
+ */
+const scopesMatch = (
+  schemaScope: SchemaElement | undefined,
+  uiScope: EditorUISchemaElement | undefined
 ) => {
-  const uiElementToMove = item.uiSchemaElement as EditorUISchemaElement;
-  // can't move the root element
-  if (!uiElementToMove.parent) {
-    return false;
+  if (!schemaScope && !uiScope) {
+    return true;
   }
-  // can't move element into itself
-  if (getHierarchy(layout).includes(uiElementToMove)) {
+  if (!(schemaScope && uiScope)) {
     return false;
   }
-  // can't move element next to itself (which would result in no change)
-  if (layout === uiElementToMove.parent) {
-    const currentIndex = (uiElementToMove.parent as EditorLayout).elements.indexOf(
-      uiElementToMove
+  return uiScope.linkedSchemaElement === schemaScope.uuid;
+};
+
+/**
+ * Returns the closest scope changing schema container
+ */
+const getScopeChangingContainer = (element: SchemaElement) => {
+  // TODO check other cases than array
+  return getArrayContainer(element);
+};
+
+export const canMoveSchemaElementTo = (
+  item: MoveUISchemaElement,
+  target: EditorUISchemaElement,
+  index: number
+) => {
+  const elementToMove = item.uiSchemaElement as EditorUISchemaElement;
+  return (
+    !isMoveRoot(elementToMove) &&
+    !isMoveIntoItself(elementToMove, target) &&
+    !isMoveNextToItself(elementToMove, target, index) &&
+    !isMovingControlsInterScopes(elementToMove, target)
+  );
+};
+
+const isMoveRoot = (elementToMove: EditorUISchemaElement) =>
+  !elementToMove.parent;
+const isMoveIntoItself = (
+  elementToMove: EditorUISchemaElement,
+  target: EditorUISchemaElement
+) => getHierarchy(target).includes(elementToMove);
+const isMoveNextToItself = (
+  elementToMove: EditorUISchemaElement,
+  target: EditorUISchemaElement,
+  index: number
+) => {
+  if (target === elementToMove.parent) {
+    const currentIndex = (target as EditorLayout).elements.indexOf(
+      elementToMove
     );
     if (currentIndex === index || currentIndex === index - 1) {
```

Comment:
why not return `currentIndex === index || currentIndex === index - 1` ?

Reply:
Same reason as in the other one. I prefer to return the specific problem as `true` and then fallback to `false` in the default path. I'm fine with changing it if you prefer.